File: java/ToolGood/Words/WordsSearchEx.java
package ToolGood.Words;

import java.util.ArrayList;
import java.util.List;

import ToolGood.Words.internals.BaseSearchEx;

public class WordsSearchEx extends BaseSearchEx{

    public List<WordsSearchResult> FindAll(String text)
    {
        List<WordsSearchResult> root = new ArrayList<WordsSearchResult>();
        int p = 0;
        int length = text.length();
        for (int i = 0; i < length; i++) {
            int t =  _dict[text.charAt(i)];
            if (t == 0) {
                p = 0;
                continue;
            }
            int next = _next[p] + t;
            boolean find = _key[next] == t;
            if (find == false && p != 0) {
                p = 0;
                next = _next[0] + t;
                find = _key[next] == t;
            }
            if (find) {
                int index = _check[next];
                if (index > 0) {
                    for (int item : _guides[index]) {
                        String key = _keywords[item];
                        WordsSearchResult r = new WordsSearchResult(key, i + 1 - key.length(), i, item);
                        root.add(r);
                    }
                }
                p = next;
            }
        }
        return root;
    }

 
    public WordsSearchResult FindFirst(String text)
    {
        int p = 0;
        int length = text.length();
        for (int i = 0; i < length; i++) {
            int t =  _dict[text.charAt(i)];
            if (t == 0) {
                p = 0;
                continue;
            }
            int next = _next[p] + t;
            if (_key[next] == t) {
                int index = _check[next];
                if (index > 0) {
                    String item = _keywords[_guides[index][0]];
                    return new WordsSearchResult(item, i + 1 - item.length(), i, _guides[index][0]);
                }
                p = next;
            } else {
                p = 0;
                next = _next[p] + t;
                if (_key[next] == t) {
                    int index = _check[next];
                    if (index > 0) {
                        String item = _keywords[_guides[index][0]];
                        return new WordsSearchResult(item, i + 1 - item.length(),i, _guides[index][0]);
                    }
                    p = next;
                }
            }
        }
        return null;
    }

    public boolean ContainsAny(String text)
    {
        int p = 0;
        for (int i = 0; i < text.length(); i++) {
            int t =  _dict[text.charAt(i)];
            if (t == 0) {
                p = 0;
                continue;
            }
            int next = _next[p] + t;
            if (_key[next] == t) {
                if (_check[next] > 0) { return true; }
                p = next;
            } else {
                p = 0;
                next = _next[p] + t;
                if (_key[next] == t) {
                    if (_check[next] > 0) { return true; }
                    p = next;
                }
            }
        }
        return false;
    }

    public String Replace(String text){
        return Replace(text,'*');
    }

    public String Replace(String text, Character replaceChar)
    {
        StringBuilder result = new StringBuilder(text);

        int p = 0;

        for (int i = 0; i < text.length(); i++) {
            int t =  _dict[text.charAt(i)];
            if (t == 0) {
                p = 0;
                continue;
            }
            int next = _next[p] + t;
            boolean find = _key[next] == t;
            if (find == false && p != 0) {
                p = 0;
                next = _next[p] + t;
                find = _key[next] == t;
            }
            if (find) {
                int index = _check[next];
                if (index > 0) {
                    int maxLength = _keywords[_guides[index][0]].length();
                    int start = i + 1 - maxLength;
                    for (int j = start; j <= i; j++) {
                        result.setCharAt(j, replaceChar);
                     }
                 }
                 p = next;
            }
        }
        return result.toString();
    }

}

File: golang/ToolGood/Words/WordsSearchEx.go
package Words

import . "./internals"

 
type WordsSearchEx struct {
	BaseSearchEx
}

func NewWordsSearchEx() *WordsSearchEx{
	return &WordsSearchEx{
	}
}

func (this *WordsSearchEx)FindAll(text string) []*WordsSearchResult  {
	list:= make([]*WordsSearchResult, 0)
	p:=0
	var i int
	for _,c := range text {
		t:=this.I_dict[c]
		if	t==0{
			p = 0;
			i++
			continue;
		}
		next := this.I_next[p] + t;
		find := this.I_key[next] == t;
		if  find == false && p != 0  {
			p = 0;
			next = this.I_next[0] + t;
			find = this.I_key[next] == t;
		}
		if  find  {
			index := this.I_check[next];
			if (index > 0) {
				for _,item:=range this.I_guides[index]{
					k:=this.I_keywords[item]
					maxLength:=len([]rune(k))
					r:= NewWordsSearchResult(k, i + 1 - maxLength, i, index);
					list=append(list,r)
				}
			}
			p = next;
		}
		i++
	}
	return list;
}
 

func (this *WordsSearchEx)FindFirst(text string) *WordsSearchResult  {
	p:=0
	var i int
	for _,c := range text {
		t:=this.I_dict[c]
		if	t==0{
			p = 0;
			i++
			continue;
		}
		next := this.I_next[p] + t;
		find := this.I_key[next] == t;
		if  find == false && p != 0  {
			p = 0;
			next = this.I_next[0] + t;
			find = this.I_key[next] == t;
		}
		if  find  {
			index := this.I_check[next];
			if (index > 0) {
				k:=this.I_keywords[this.I_guides[index][0]]
				maxLength:=len([]rune(k))
				return NewWordsSearchResult(k, i + 1 - maxLength, i, index);
			}
			p = next;
		}
		i++
	}
	return nil;
}
 

func (this *WordsSearchEx)ContainsAny(text string) bool  {
	p:=0
	for _,c := range text {
		t:=this.I_dict[c]
		if	t==0{
			p = 0;
			continue;
		}
		next := this.I_next[p] + t;

		if this.I_key[next] == t {
			if  this.I_check[next] > 0  { 
				return true;
			}
			p = next;
		} else {
			p = 0;
			next = this.I_next[p] + t;
			if (this.I_key[next] == t) {
				if (this.I_check[next] > 0) {
					 return true;
				}
				p = next;
			}
		}
	}
	return false;
}
 

func (this *WordsSearchEx)Replace(text string, replaceChar rune) string  {
	result:= []rune (text)
	p:=0
	var i int
	for _,c := range text {
		t:=this.I_dict[c]
		if	t==0{
			p = 0;
			i++
			continue;
		}
		next := this.I_next[p] + t;
		find := this.I_key[next] == t;
		if  find == false && p != 0  {
			p = 0;
			next = this.I_next[0] + t;
			find = this.I_key[next] == t;
		}
		if  find  {
			index := this.I_check[next];
			if (index > 0) {
				r:=this.I_keywords[this.I_guides[index][0]]
				maxLength:=len([]rune(r))
				start:= i + 1 - maxLength
				for j := start; j <= i; j++{
					result[j]=replaceChar
				} 
			}
			p = next;
		}
		i++
	}
	return string (result) 
}
